### ISSUE TITLE
Add resources to create NSX-T security tags and manage assignment

### ISSUE THREAD

**Issue by jpbuecken** [vmware/terraform-provider-vcd#459]
Hi,

could you please add support to create security tags and manage it members:

This is the corresponding vCD documentation:

https://docs.vmware.com/en/vCloud-Director/9.5/com.vmware.vcloud.tenantportal.doc/GUID-681D5E9C-8DD1-485C-A007-DFD2C1AABB70.html

I'm not sure about the best way to implement management of membership. AWS, OCI and Openstack implement this as part of the port/network interface or of the VM/instance resource.
But vCD follow another approach in its GUI. Straight forward is to configure membership in the resource which create it.
Or you implement an attachment resource which attach a security tag to a member (VM, Network, ....)
All three ways will have its advantage and disadvantage.

### Terraform Version
ALL

### Affected Resource(s)
vcd_security_tag (suggestion)
vcd_security_tag_assignment (suggestion)

### Terraform Configuration Files
```hcl
resource "vcd_vapp_vm" "web1" {
[...]
}
resource "vcd_vapp_vm" "web2" {
[...]
}
resource "vcd_security_tag" "webrules" {
   name = "applicationset1"
   description = "Rules for application 1"
}

resource "vcd_security_tag_assignment" "webassign" {
   tagid = vcd_security_tag.webrules.id
   # array of instance ids
   members = [ vcd_vapp_vm.web1.id, vcd_vapp_vm.web2.id ]
}
```

**Comment by Didainius** [comment]
Hi,
What is your end goal? Is it dynamic security groups for fire-walling?

**Comment by jpbuecken** [comment]
Our product catalog allows you to order a server with an "intended use".
e.g. if you choose SAP or Webserver we can extend the firewall ruleset for further monitoring as a day one operation during installation.

Content and audit conform documentation of rules in the security group is done by team A. 
OS management is team B. 
The security tag feature allows team A to provide pre-defined rulesets which team B can use.

**Comment by storbel** [comment]
@Didainius 
IS there a way to create dynamic groups with VCD? 
sound that "vcd_security_tag" is not yet implemented with vcd 3.51, have any clue when this will be implemented ?

**Comment by Didainius** [comment]
@storbel ,
We're heads down for VDC group and general NSX-T support so these are definitely priorities however I can't yet promise if security tags will appear in the next release.

**Comment by Didainius** [comment]
An update on this.

@jpbuecken , @storbel  - you had ideas so I'd like to share where we're going and possibly get your feedback:

This issue was originally raised for NSX-V Security Tags, but as NSX-V is being phased out and VCD 10.3+ got support for NSX-T Security tags, we will work on adding NSX-T Security Tag support here for VCD 10.3+.

Now as for the design of this resource in NSX-T API I do believe this is a single resource (we have yet to decide between `vcd_security_tag` or `vcd_nsxt_security_tag`) due to the fact that API uses the same endpoint for managing tags and descriptions (`cloudapi/1.0.0/securityTags/tag`). There is also another endpoint for handling tags per VM `cloudapi/1.0.0/securityTags/vm/{id}`, but I think we could avoid it for now (at least on the initial version).

Sample payload is quite straight forward for endpoint `cloudapi/1.0.0/securityTags/tag` so that would be a smallish resource. Once this is in place, we can think of Dynamic Security Groups (queries by  @storbel ) which I moved to a separate issue #787.
```json
{
                "entities": [
                                "urn:vcloud:vm:f28d580e-b307-4980-a3c7-ef5b9bf560ad",
                                "urn:vcloud:vm:b438cf33-fbb3-4b64-88b0-1d1d09904014"
                ],
                "tag": "tag1"
}
```

Screenshot from API docs:
<img width="350" alt="image" src="https://user-images.githubusercontent.com/15804230/154652030-9e71bd78-256d-4d50-ae90-b1e807ffbd75.png">

**Comment by jpbuecken** [comment]
Hello, nice you ask for feedback. I had a meeting with our NSX team today.

Our use case is: The NSX team does the firewall handling/network security and prepares security groups. Those security groups are considered pre-approved to use by a OperatingSystem (OS) admin.

Our OS team has a per VM view and our current terraform project design fits this.

This separation of responsibility would better fit if the OS team can manage tags in vcd_vapp_vm resource, e.g. via

```
There is also another endpoint for handling tags per VM cloudapi/1.0.0/securityTags/vm/{id}, [...]
```

So our use case needs the vm securityTags API in vcd_vapp_vm. 


Independent of this:

Will entities be optional?

I found 3 ways to add security tags via UI: 
Networking -> Security Tags (via UI, one entitiy is mandatory)
Data Center -> (Choose a VDC) -> Security -> Configure Services of a Security Service -> Security Tags: This Tag can be empty, entities are optional!
Data Center -> (Choose a VDC) -> Virtual Machines -> (Choose Details of a VM) -> Security Tags

Since we want to split responsibility between creation of a tag, and assignment of a tag to VMs, entities should be optional if possible:

e.g.
by NSX team, it defines the tag:
```
resource "vcd_nsxt_security_tag" "tag1" {
                "entities": null
                "tag": "tag1"
}
```` 
by OS team, it uses the pre-defined tag
```
resource "vcd_vapp_vm" "vm" {
[...]
                "securitytags": "[tag1]"
[...]
}
```

But maybe there are VCD API restrictions that I'm not aware of.

**Comment by Didainius** [comment]
Just mentioning the WIP PR here as we're working on it -> https://github.com/vmware/terraform-provider-vcd/pull/845

**Comment by storbel** [comment]
@Didainius @mikeletux  cloudapi/1.0.0/securityTags/tag and cloudapi/1.0.0/securityTags/vm/{id} are representing two different  resources, in a standard workflow, you define a new securityTag on the vCloud/NsxT first, then whenever you create a new VM you give it a securityTag, that's why you have both API endpoints.

Suggestion: 
. Define a security tag on vCloud tenant: 
resource "vcd_nsxt_security_tag" "tag1" {
                "name" =  "tag1"
}
. Define security tags under a VM, two ways : 

1. When defining a new VM : 
resource "vcd_vapp_vm" "vm" {
[...]
                "security_tags" =  ["tag1","tag2",...]
[...]
}
2. Add tag to existing VM:
 
resource "vcd_vm_nsxt_security_tag" "mapping_vm1_stags" {
                vm_name =  "vm1"
                security_tag_names = ["tag1","tag2"]
}

**Comment by Didainius** [comment]
@storbel,
Timely input! @mikeletux  is working on tags now (after that we will tackle dynamic security groups #787). 
There is a WIP PR https://github.com/vmware/terraform-provider-vcd/pull/845 to introduce that with a current mockup. 
* Any thoughts on the current proposed layout? 
* Do you think you'd rather use the approach you have listed just above here?
  * Any use case for the new resource `vcd_vm_nsxt_security_tag` as opposed to just using "Update" operation for `vcd_vapp_vm`?

**Comment by jpbuecken** [comment]
The process @storbel describes in https://github.com/vmware/terraform-provider-vcd/issues/459#issuecomment-1121931417 looks similar to my use case/comment https://github.com/vmware/terraform-provider-vcd/issues/459#issuecomment-1047724974 from 22 Feb.

Create the tag. Use the tag in VM. 
And in our company, those tasks are done two different teams (NSX experts and OS operating team).
That is the use case for vcd_vm_nsxt_security_tag.

I need to look into it again, but we use roles in VCD.
So maybe the OS team does not have permission to create security tag. Without NSX-T (firewall) rules, the security tag is not very useful anyway...

**Comment by jpbuecken** [comment]
Nice work, I forwarded it to our VCD / NSX team instantly . I have to admit I wasn't aware of dynamic SG myself, but I will discuss with our team if they lift the limitations described by @mikeletux  for our use cases.
